docs/Architecture.md

```diff
@@ -0,0 +1,502 @@
+Babylon and the BSN over IBC. The `babylon-contract` contract is responsible for
+handling the IBC messages and relaying them to the appropriate contracts on the
+BSN side.
+It also deals with contract's instantiation, and with storing the information of
+a transfer channel between the BSN and Babylon, for opt-in rewards transfer in
+the form of the BSN's native token, to Babylon.
+
+Another important contract on the BSN side is the `btc-finality` contract,
+which handles the information about signatures and public randomness sent by
+the finality providers operating on the BSN. It also deals with vote tallying,
+and defines the finalisation status of the blocks on the BSN.
+
+Finally, the `btc-light-client` contract is responsible for maintaining the
+light client state of the Bitcoin network on the BSN. It is used to verify
+the delegations and timestamping sent through Babylon, and to ensure that the
+BSN is in sync with the Bitcoin network.
+
+## Detailed Interfaces
+
+This section provides a detailed overview of the interfaces provided by the
+different contracts involved in the Babylon Staking integration.
+It only describes the interfaces that are relevant for the Babylon Staking
+integration, and does not cover the full functionality of the contracts. It also
+does not cover the different queries that the contracts provide, which can be
+found in the contract's documentation.
+
+### Messages
+
+There are four main message types that are used in the Babylon Staking
+integration.
+
+- IBC messages, which are used to communicate between the Babylon and BSN sides.
+- Execute messages, which are used to execute actions on the contracts.
+- Sudo messages, which are used by the `babylon` module which is part of
+  babylon-sdk to interact with the Babylon contracts; typically during begin /
+  end block processing.
+- Custom messages, which are used by the contracts to interact with the
+  `babylon` module and execute privileged actions through the Cosmos SDK.
+
+There are also instantiation messages, which are used to instantiate the
+contracts on the BSN side, and to set up the necessary channels for IBC
+communication between the Babylon and BSN sides.
+These messages are not covered in this document, as they are part of the
+contract's instantiation process and are typically handled by the integrator
+during the chain setup.
+
+There are also queries, that can be used to query the state of the contracts,
+that are not covered in this document either.
+The queries can be found in the contract's documentation, and are used to query
+the state of the contracts and the information related to the delegations,
+finality providers, and other information related to the Babylon Staking
+integration.
+
+### Messages Flow
+
+The following diagram outlines the flow of messages between the Babylon
+and BSN sides, as well as the contracts involved, and the babylon-sdk `babylon`
+module.
+```mermaid
```

Comment:
This graph is hard to read, if we keep it, I suggest we simplify.

Reply:
The graph is hard to read, because the architecture and the messages flow is complex.

I'm not sure we can simplify this. Perhaps ditch the diagram entirely?

Reply:
I propose to move this diagram to a `docs/MESSAGES.md` file. And reference it from the architecture doc.